npm/pkg/controlplane/controllers/v2/networkPolicyController.go

```diff
@@ -0,0 +1,339 @@
+}
+
+// syncAddAndUpdateNetPol handles a new network policy or an updated network policy object triggered by add and update events
+func (c *NetworkPolicyController) syncAddAndUpdateNetPol(netPolObj *networkingv1.NetworkPolicy) error {
+	prometheusTimer := metrics.StartNewTimer()
+	defer metrics.RecordPolicyExecTime(prometheusTimer) // record execution time regardless of failure
+
+	var err error
+	netpolKey, err := cache.MetaNamespaceKeyFunc(netPolObj)
+	if err != nil {
+		return fmt.Errorf("[syncAddAndUpdateNetPol] Error: while running MetaNamespaceKeyFunc err: %w", err)
+	}
+
+	// Start reconciling loop to eventually meet cached states against the desired states from network policy.
+	// #1 If a new network policy is created, the network policy is not in RawNPMap.
+	// start translating policy and install translated ipset and iptables rules into kernel
+	// #2 If a network policy with <ns>-<netpol namespace>-<netpol name> is applied before and two network policy are the same object (same UID),
+	// first delete the applied network policy, then start translating policy and install translated ipset and iptables rules into kernel
+	// #3 If a network policy with <ns>-<netpol namespace>-<netpol name> is applied before and two network policy are the different object (different UID) due to missing some events for the old object
+	// first delete the applied network policy, then start translating policy and install translated ipset and iptables rules into kernel
+	// To deal with all three cases, we first delete network policy if possible, then install translated rules into kernel.
+	// (TODO): can optimize logic more to reduce computations. For example, apply only difference if possible like podController
+
+	// Do not need to clean up default Azure NPM chain in deleteNetworkPolicy function, if network policy object is applied soon.
+	// So, avoid extra overhead to install default Azure NPM chain in initializeDefaultAzureNpmChain function.
+	// To achieve it, use flag unSafeToCleanUpAzureNpmChain to indicate that the default Azure NPM chain cannot be deleted.
+	// delete existing network policy
+	err = c.cleanUpNetworkPolicy(netpolKey)
+	if err != nil {
+		return fmt.Errorf("[syncAddAndUpdateNetPol] Error: failed to deleteNetworkPolicy due to %w", err)
+	}
+
+	// install translated rules into kernel
+	npmNetPolObj := translation.TranslatePolicy(netPolObj)
+
+	fmt.Printf("%+v", npmNetPolObj)
+
+	// install translated rules into Dataplane
+	err = c.dp.AddPolicy(npmNetPolObj)
+	if err != nil {
+		return fmt.Errorf("[syncAddAndUpdateNetPol] Error: failed to install translated NPMNetworkPolicy into Dataplane due to %w", err)
+	}
+
+	// Cache network object first before applying ipsets and iptables.
+	// If error happens while applying ipsets and iptables,
+	// the key is re-queued in workqueue and process this function again, which eventually meets desired states of network policy
+	c.rawNpMap[netpolKey] = netPolObj
+	metrics.IncNumPolicies()
```

Comment:
should we track policy metrics in the controller or DP? IPSet metrics must be tracked in the DP since the logic is only feasible in the ipset manager when we apply sets on need in the kernel

Reply:
I am okay with either approaches, since the object is just one and logic is simple if we do it in controller, since we will only increment if DP sends a no error signal. In DP we can have more robust metrics to determine, how many endpoints have a policy, how many rules etc .

Reply:
In DP policy may not make sense, but rules do. so we should have a metric for rules in DP, wdyt ?

Reply:
Makes sense to have rules in DP. Good with the code as is, but if the number of policies is the only metric that the control plane could take care of, then perhaps we just take care of that metric in DP only for simplicity